src/rust/engine/src/nodes.rs

```diff
@@ -942,13 +942,20 @@ impl Task {
               .ok_or_else(|| match get.declared_subject {
                 Some(ty) if externs::is_union(ty) => {
                   let value = externs::get_value_from_type_id(ty);
-                  let description = externs::project_str(&value, "union_description");
-                  throw(&format!(
-                    "Type {} is not a member of the {} @union (\"{}\")",
-                    get.subject.type_id(),
-                    ty,
-                    description
-                  ))
+                  match externs::call_method(
+                    &value,
+                    "non_member_error_message",
+                    &[externs::val_for(&get.subject)],
+                  ) {
+                    Ok(err_msg) => throw(&externs::val_to_str(&err_msg)),
+                    // If the non_member_error_message() call failed for any reason,
```

Comment:
Is it possible we could add in a check for whether the union contains the `non_member_error_message` attr? If for example my implementation of `non_member_error_message` happens to be throwing, I feel like I wouldn't want to see the same message as if there wasn't any matching `UnionRule`. If we don't currently have an extern method to check whether an attribute exists, we could potentially have the default `non_member_error_message` just return None, and then check whether the result of the `call_method()` is equal to `externs::none` (see https://github.com/pantsbuild/pants/blob/8162b77dd29c1d4216861466df0f8ed8b956df0f/src/rust/engine/engine_cffi/src/lib.rs#L785)

Reply:
Not sure what you're asking for? The union *always* implements `non_member_error_message`.  If the underlying class (the one decorated by @union) doesn't define it then the union creates a default one. 

This extra check is just for completeness, in case the underlying custom `non_member_error_message` throws for some unlikely reason. And this message will be different from the default one (the default one contains the docstring). I'm not sure there's more to do here.

Reply:
I'm concerned about:
```python
@union
class TestTarget:

  @staticmethod
  def non_member_error_message(subject):
    if hasattr(subject, 'address'):
      return f'{subject.address.reference()} is not a testable target.'
    else:
      raise Exception(f"something awful happened!!! {subject}")
```
I think right now the current implementation would swallow the exception and say `Type X is not a member of the TestTarget @union`? That seems like a potential source of confusion, since that error message does not describe the actual error that's occurring.

Reply:
Well, keep in mind that the fallback error message is correct: The type is indeed not a member of the union. In the unlikely event that we threw while generating the more specific error message, it seems like we wouldn't want to send the user down that rabbit hole. That seems less actionable than the more important issue of the union mismatch?